src/mumble/GlobalShortcut_win.cpp

```diff
@@ -719,211 +421,170 @@ void GlobalShortcutWin::timeTicked() {
-	// the mouse in Windows becomes laggy. It makes the whole computer feel
-	// like it has locked up -- because input "stops".
-	//
-	// As explained above, initializing the hooks here yields a much superior
-	// experience, where this initialization has no observable effect on the
-	// behavior of the system's mouse input.
-	if (bHook && !hhMouse && !hhKeyboard) {
-		HMODULE hSelf;
-		GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
-						  reinterpret_cast< LPCTSTR >(&HookKeyboard), &hSelf);
-		hhMouse    = SetWindowsHookEx(WH_MOUSE_LL, HookMouse, hSelf, 0);
-		hhKeyboard = SetWindowsHookEx(WH_KEYBOARD_LL, HookKeyboard, hSelf, 0);
-	}
-}
+			handleButton(ql, m_gkey->isMouseButtonPressed(btn));
+		}
 
-GlobalShortcutWin::ButtonInfo GlobalShortcutWin::buttonInfo(const QVariant &v) {
-	GlobalShortcutWin *gsw = static_cast< GlobalShortcutWin * >(GlobalShortcutEngine::engine);
+		for (uint8_t mode = GKEY_MIN_KEYBOARD_MODE; mode <= GKEY_MAX_KEYBOARD_MODE; mode++) {
+			for (uint8_t key = GKEY_MIN_KEYBOARD_BUTTON; key <= GKEY_MAX_KEYBOARD_BUTTON; key++) {
+				const QList< QVariant > ql{ QVariant::fromValue(DeviceTypes::GKEY_KEYBOARD),
+											// Store the key and mode in one int
+											// bit 0..15: mode, bit 16..31: key
+											key | (mode << 16) };
 
-	const QList< QVariant > &sublist = v.toList();
-	if (sublist.count() != 2) {
-		return ButtonInfo();
+				handleButton(ql, m_gkey->isKeyboardGkeyPressed(key, mode));
+			}
+		}
 	}
+#endif
+}
 
-	bool ok    = false;
-	DWORD type = sublist.at(0).toUInt(&ok);
-	QUuid guid(sublist.at(1).toString());
+GlobalShortcutWin::ButtonInfo GlobalShortcutWin::buttonInfo(const QVariant &button) {
+	auto gsw = static_cast< GlobalShortcutWin * >(GlobalShortcutEngine::engine);
 
-	if (guid.isNull() || (!ok)) {
-		return ButtonInfo();
-	}
+	const auto &sublist   = button.toList();
+	const auto deviceType = sublist.at(0).value< DeviceTypes >();
 
 	ButtonInfo info;
-	info.device = guid.toString();
 
-#ifdef USE_GKEY
-	if (Global::get().s.bEnableGKey && gkey && gkey->isValid()) {
-		bool isGKey = false;
-		if (guid == GKeyLibrary::quMouse) {
-			isGKey    = true;
-			info.name = gkey->getMouseButtonString(type);
-		} else if (guid == GKeyLibrary::quKeyboard) {
-			isGKey    = true;
-			info.name = gkey->getKeyboardGkeyString(type & 0xFFFF, type >> 16);
-		}
-		if (isGKey) {
-			// Example output:
-			// "Logitech G-keys"
-			// "G6/M1"
-			info.device       = QLatin1String("Logitech G-keys");
-			info.devicePrefix = QLatin1String("GKey:");
-			return info;
-		}
-	}
-#endif
+	switch (deviceType) {
+		case DeviceTypes::KEYBOARD: {
+			info.device         = tr("Keyboard");
+			const auto scanCode = sublist.at(1).toUInt();
+			const auto isE0     = sublist.at(2).toBool();
+
+			wchar_t buffer[MAX_PATH];
+			if (GetKeyNameText((scanCode << 16) | (isE0 << 24), buffer, sizeof(buffer) / sizeof(wchar_t))) {
```

Comment:
Also what is the bit shifting needed for?

Reply:
From https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getkeynametexta:

![](https://user-images.githubusercontent.com/5897523/116003020-9c722500-a5fc-11eb-93ae-4e578a12a2fd.png)